src/MongoDB.Driver.Core/Core/Connections/TcpStreamFactory.cs

```diff
@@ -250,13 +250,27 @@ private Socket CreateSocket(EndPoint endPoint)
             }
 
             var socket = new Socket(addressFamily, SocketType.Stream, ProtocolType.Tcp);
-            var keepAliveValues = new KeepAliveValues
+            try
             {
-                OnOff = 1,
-                KeepAliveTime = 300000, // 300 seconds in milliseconds
-                KeepAliveInterval = 10000 // 10 seconds in milliseconds
-            };
-            socket.IOControl(IOControlCode.KeepAliveValues, keepAliveValues.ToBytes(), null);
+                var keepAliveValues = new KeepAliveValues
+                {
+                    OnOff = 1,
+                    KeepAliveTime = 300000, // 300 seconds in milliseconds
+                    KeepAliveInterval = 10000 // 10 seconds in milliseconds
+                };
+                socket.IOControl(IOControlCode.KeepAliveValues, keepAliveValues.ToBytes(), null);
+            }
+            catch (PlatformNotSupportedException)
+            {
+                try
+                {
+                    socket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.KeepAlive, true);
+                }
+                catch (PlatformNotSupportedException)
```

Comment:
I'm inclined to remove this catch block if Vincent can confirm it's not hit on Linux or OS X. If it is hit, we should document that keep alive is not supported at all on Linux/OS X.  CC @vincentkam.

Reply:
I was thinking more about the fact that there are many platforms and we don't know what is and is not supported on each platform.

Isn't it better to make a best case effort to turn on KeepAlive.

It's not enough to document that keepAlive is not supported. If we throw here the driver won't be able to connect at all.

Reply:
> I was thinking more about the fact that there are many platforms and we don't know what is and is not supported on each platform.

That logic applies to every platform method we call.  Given the current state of the .NET ecosystem, we don't really know unless we test.

> Isn't it better to make a best case effort to turn on KeepAlive.

Yes, but we can prove that SetSocketOption works on supported platforms by testing on those platforms (in this case manually). 

> It's not enough to document that keepAlive is not supported. If we throw here the driver won't be able to connect at all.

Agreed on this point.  We can't throw, so we should only remove the catch block if we can prove that SetSocketOption is supported on all platforms that don't support the IOControl method.  We do want to document it though, so I'd like to know if we enter that catch block on any supported platform.

But if you still want to leave this in due to our current lack of platform test coverage, I'm ok with it.

Reply:
In theory we can test SetSocketOption on all supported platforms, but there are too many of them for that to be an appealing option. And there may be new platforms in the future.

I agree that it would be nice to know exactly which platforms SetSocketOption is supported on, specifically when called with the KeepAlive option. But that's a lot of work, and more than we want to do, at least now.